explorer/interpreter/type_checker.cpp

```diff
@@ -3029,6 +3134,21 @@ auto TypeChecker::TypeCheckExp(Nonnull<Expression*> e,
 
       auto& self = where.self_binding();
       ConstraintTypeBuilder builder(arena_, &self);
+      ConstraintTypeBuilder::ConstraintsInScopeTracker constraint_tracker;
+
+      // Keep track of the builder so that we can look up its rewrites while
+      // processing later constraints.
+      struct TrackPartialWhereBuilderRAII {
+        TrackPartialWhereBuilderRAII(TypeChecker& type_checker,
+                                     ConstraintTypeBuilder& builder)
+            : type_checker_(type_checker) {
+          type_checker_.partial_where_expressions_.push_back(&builder);
+        }
+        ~TrackPartialWhereBuilderRAII() {
+          type_checker_.partial_where_expressions_.pop_back();
+        }
+        TypeChecker& type_checker_;
+      } partial_where_tracker(*this, builder);
```

Comment:
```suggestion
partial_where_expressions_.push_back(builder);
auto partial_where_popper = llvm::make_scope_exit([]{partial_where_expressions_.pop_back();});
```

How much is the struct helping here?

Reply:
Thanks, I had forgotten LLVM provided this.